irods/test/data_obj_test.py

```diff
@@ -2362,6 +2363,79 @@ def _update(n):
 
         self.assertEqual(pbar.percent_done(), 100)
 
+    def test_replica_truncate_related_errors__issue_534(self):
+        sess = self.sess
+        data_objs = self.sess.data_objects
+        truncated_size = 16
+
+        try:
+            # Set path for a new, as yet nonexisting data object.
+            data_path = '{}/{}'.format(
+                helpers.home_collection(sess),
+                unique_name(my_function_name(), datetime.now(), truncated_size) + "issue_534_errors")
+            self.assertFalse (data_objs.exists(data_path), "Data object should not yet exist.")
+
+            # Assert appropriate error is raised for the nonexisting object.
+            with self.assertRaises(ex.OBJ_PATH_DOES_NOT_EXIST):
+                data_objs.replica_truncate(data_path, truncated_size)
+
+            d = data_objs.create(data_path)
+
+            # Assert appropriate error is raised for an existing object without a replica on the given resource.
+            with self.create_simple_resc() as newResc:
+                with self.assertRaises(ex.SYS_REPLICA_INACCESSIBLE):
+                    d.replica_truncate(truncated_size, **{kw.RESC_NAME_KW: newResc})
+
+            # Assert appropriate error is raised for an invalid (negative) size argument.
+            try:
+                data_objs.replica_truncate(data_path, -truncated_size)
```

Comment:
Are you able to capture the output when these exceptions are thrown, or do we just get the error code?

If it's possible to get the JSON output in the error case, let's try to capture it and... well, we don't need to assert the contents per se, but we can make sure that it's valid JSON, perhaps. Thoughts?

Reply:
Tried stepping through, and it appears the response is parsed as an error, rather than  as a routine response with JSON output expected.  SHould the PRC be looking for the JSON output in addition to an error code?

Reply:
Gotcha. The JSON output in error cases would provide an error message from the server that could be useful for applications built on top of this one. I guess PRC is more relying on the rError stack for things like that, so it can't be helped. Let's just let it ride because it's fine as is. Resolving.

Reply:
Ok.  Just wondering, then, will the custom of using an rError stack soon be in disfavor, with the useful information coming back via the JSON message response instead?  Is that perhaps the "new way"?

Reply:
> Ok. Just wondering, then, will the custom of using an rError stack soon be in disfavor, with the useful information coming back via the JSON message response instead? Is that perhaps the "new way"?

I redirect this question to @korydraughn. Many of the newer APIs are using JSON output, but not all. I don't think we've settled on a direction yet. It does seem (to me) that rError usage has fallen out of favor.

Reply:
I wouldn't say one way is favored over the other. It depends on the design of the API.

iRODS provides multiple ways of doing things. For the replica truncate API, returning JSON via an out parameter made sense. A different API may pass things back via the rError stack and there's nothing wrong with that _(even if it's not preferred)_.

The important thing is that we continue to document expected usage and behavior so users don't get confused and lose confidence in the interface.

Reply:
Ah, ok.  I may need to bake into the client some way of getting to the JSON out parameter in the case of error.  Presently, the irods.message interface throws an exception rather than allowing application access to the return message that causes it.

Reply:
Should we push this off for a later release, or do it now? Asking because I resolved the conversation but it looks like we're looking into more work... Can unresolve if that seems helpful.

Reply:
Let's punt on replica truncate and get the PRC released.

The 2.1.0 milestone has 25 closed issues which is plenty.

Reply:
After some discussion, we've decided to open an issue for 3.0 to investigate allowing the out parameter to be captured on error and allow this PR to be merged.

Reply:
The new issue is #606. There is a PR to add the functionality in #607. I think this can be resolved.

Reply:
Yes, agree, ... resolving.